src/drivers/intel/ssp/ssp.c

```diff
@@ -621,73 +726,21 @@ static int ssp_set_config(struct dai *dai, struct ipc_config_dai *common_config,
 static int ssp_pre_start(struct dai *dai)
 {
 	struct ssp_pdata *ssp = dai_get_drvdata(dai);
-	struct sof_ipc_dai_config *config = &ssp->config;
-	uint32_t sscr0;
-	uint32_t mdiv;
-	bool need_ecs = false;
-
 	int ret = 0;
 
 	dai_info(dai, "ssp_pre_start()");
 
-	/* SSP active means bclk already configured. */
-	if (ssp->state[SOF_IPC_STREAM_PLAYBACK] == COMP_STATE_ACTIVE ||
-	    ssp->state[SOF_IPC_STREAM_CAPTURE] == COMP_STATE_ACTIVE)
-		return 0;
+	/*
+	 * We will test if mclk/bclk is configured in
+	 * ssp_mclk/bclk_prepare_enable/disable functions
+	 */
 
 	/* MCLK config */
-	ret = mn_set_mclk(config->ssp.mclk_id, config->ssp.mclk_rate);
-	if (ret < 0) {
-		dai_err(dai, "invalid mclk_rate = %d for mclk_id = %d",
-			config->ssp.mclk_rate, config->ssp.mclk_id);
-		goto out;
-	}
+	ret = ssp_mclk_prepare_enable(dai);
+	if (ret < 0)
+		return ret;
 
-	sscr0 = ssp_read(dai, SSCR0);
-
-#if CONFIG_INTEL_MN
-	/* BCLK config */
-	ret = mn_set_bclk(config->dai_index, config->ssp.bclk_rate,
-			  &mdiv, &need_ecs);
-	if (ret < 0) {
-		dai_err(dai, "invalid bclk_rate = %d for dai_index = %d",
-			config->ssp.bclk_rate, config->dai_index);
-		goto out;
-	}
-#else
-	if (ssp_freq[SSP_DEFAULT_IDX].freq % config->ssp.bclk_rate != 0) {
-		dai_err(dai, "invalid bclk_rate = %d for dai_index = %d",
-			config->ssp.bclk_rate, config->dai_index);
-		goto out;
-	}
-
-	mdiv = ssp_freq[SSP_DEFAULT_IDX].freq / config->ssp.bclk_rate;
-#endif
-
-	if (need_ecs)
-		sscr0 |= SSCR0_ECS;
-
-	/* clock divisor is SCR + 1 */
-	mdiv -= 1;
-
-	/* divisor must be within SCR range */
-	if (mdiv > (SSCR0_SCR_MASK >> 8)) {
-		dai_err(dai, "ssp_pre_start(): divisor %d is not within SCR range",
-			mdiv);
-		ret = -EINVAL;
-		goto out;
-	}
-
-	/* set the SCR divisor */
-	sscr0 &= ~SSCR0_SCR_MASK;
-	sscr0 |= SSCR0_SCR(mdiv);
-
-	ssp_write(dai, SSCR0, sscr0);
-
-	dai_info(dai, "ssp_set_config(), sscr0 = 0x%08x", sscr0);
-out:
-
-	return ret;
+	return ssp_bclk_prepare_enable(dai);
```

Comment:
It's a good question. The two clocks are independent so technically there is no real dependency.
I think it would make sense if the error handling went back to no clock enabled at all.

@bardliao what do you think? I don't want to change this code you wrote without asking :-)

Reply:
That is a complicated case. IMO, I don't think we should simply disable MCLK without any condition. The main reason is that MCLK may be used by another SSP port. Besides, we didn't do it in the existing code.

Reply:
no change.